File: 2022/day07.py
import time
from collections import defaultdict
from functools import partial
from operator import ge, le

start = time.time()
with open("../inputs/2022/day07.input") as f:
    inp = f.read().split("\n")


def traverse(inp: list) -> dict[tuple, int]:
    loc = ()
    tree = defaultdict(int)
    for line in inp:
        if line[:3] == "$ c":
            p = line.split(" ")[-1]
            if p == "..":
                loc = loc[:-1]
            else:
                loc = loc + (p,)
        elif line[0] not in {"d", "$"}:
            size = int(line.split(" ")[0])
            for i in range(len(loc) + 1):
                tree[loc[:i]] += size
    return tree


def part1(inp: list) -> int:
    return sum(filter(partial(ge, 100_000), traverse(inp).values()))


def part2(inp: list) -> int:
    tree = traverse(inp)
    capacity = 70000000
    want = 30000000
    size = tree[("/",)]
    need = want - (capacity - size)
    return min(filter(partial(le, need), tree.values()))


print(
    f"Day 07     {part1(inp):<14} {part2(inp):<14} {(time.time() - start)*1e3:>11.2f}"
)

File: rust-2022/src/day07.rs
use std::iter::Peekable;

pub fn read(inp: &[u8]) -> (Vec<u32>, u32) {
    let mut v = Vec::with_capacity(512);
    let it = &inp[12..];
    let size = traverse(&mut it.split(|&c| c == b'\n').peekable(), &mut v);
    (v, size)
}

fn traverse<'a>(lines: &mut Peekable<impl Iterator<Item = &'a [u8]>>, out: &mut Vec<u32>) -> u32 {
    let mut size = 0;
    while let Some(&line) = lines.peek() {
        match &line[0] {
            b'$' => break,
            b'd' => (),
            _ => {
                size += line
                    .iter()
                    .take_while(|&&c| c != b' ')
                    .fold(0u32, |acc, &c| acc * 10 + (c - b'0') as u32)
            }
        }
        lines.next();
    }
    while let Some(&line) = lines.peek() {
        if let Some(b'.') = line.get(5) {
            break;
        }
        // Skip $ cd foo
        lines.next();
        // Skip $ ls
        lines.next();
        // Read sub-size
        size += traverse(lines, out);
    }
    lines.next();
    out.push(size);
    size
}

pub fn part1(inp: &(Vec<u32>, u32)) -> u32 {
    inp.0.iter().filter(|&&x| x <= 100_000).sum()
}

pub fn part2(inp: &(Vec<u32>, u32)) -> u32 {
    let (capacity, want) = (70000000, 30000000);
    *inp.0.iter().filter(|&&x| x >= want - (capacity - inp.1)).min().unwrap_or(&0)
}

pub fn run() -> (String, String) {
    let inp = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day07.input"));
    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}
